src/t_zset.c

```diff
@@ -1161,19 +1161,20 @@ unsigned char *zzlDeleteRangeByScore(unsigned char *zl, zrangespec *range, unsig
     eptr = zzlFirstInRange(zl,range);
     if (eptr == NULL) return zl;
 
+    unsigned char *first_ele = eptr;
     /* When the tail of the listpack is deleted, eptr will be NULL. */
     while (eptr && (sptr = lpNext(zl,eptr)) != NULL) {
         score = zzlGetScore(sptr);
         if (zslValueLteMax(score,range)) {
-            /* Delete both the element and the score. */
-            zl = lpDeleteRangeWithEntry(zl,&eptr,2);
             num++;
+            eptr = lpNext(zl,sptr);
         } else {
             /* No longer in range. */
             break;
         }
     }
-
+    if (num)
+        zl = lpDeleteRangeWithEntry(zl, &first_ele, 2 * num);
```

Comment:
After the above loop, we can get the range's tail, which is `eptr` (if NULL it means it's the last one in listpack). I think we could implement another function `lpDeleteRangeWithEntryPtr(unsigned char *lp, unsigned char **first, unsigned char **tail)` to avoid the traversal search for tail within `lpDeleteRangeWithEntry`.

Reply:
In this case we also need to pass the entry number between first and tail, otherwise we can't update the entry numbers, but that also make this method a little odd.

Reply:
I find we have the same problem in `zpopmin/max` with count. In `zpopmin/max`, we also get the start position, end position and the number elements to be removed. Shall I add a new function like `lpDeleteRangeWithEntryPtr` to solve both of them?

Reply:
@lyq2333 we can go ahead with this way, we can compromise on it if it really brings a boost.

Reply:
@sundb @soloestoy I add the new function `lpDeleteRangeWithEntryPtr` and make a test. The result is as follow. It improve about 6% based on the origin branch(not unstable).
```
Summary:
  throughput summary: 68101.34 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.658     0.160     0.639     0.871     0.927     3.063
```

Reply:
I also make a test about `zpopmin`. I put the result in the top comment. The performance improvement is about 8% because we costs many time on reply.